File: trie/212-word-search-ii/WordSearchII.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.word = None

class Solution:
    def findWords(self, board, words):
        root = self.build_trie(words)
        res = []

        for i in range(len(board)):
            for j in range(len(board[0])):
                self.dfs(board, i, j, root, res)
        return res

    def build_trie(self, words):
        root = TrieNode()
        for word in words:
            node = root
            for ch in word:
                if ch not in node.children:
                    node.children[ch] = TrieNode()
                node = node.children[ch]
            node.word = word
        return root

    def dfs(self, board, i, j, node, res):
        ch = board[i][j]
        if ch == '#' or ch not in node.children:
            return
        node = node.children[ch]
        if node.word:
            res.append(node.word)
            node.word = None  # prevent duplicates

        board[i][j] = '#'
        for x, y in [(1,0),(-1,0),(0,1),(0,-1)]:
            nx, ny = i + x, j + y
            if 0 <= nx < len(board) and 0 <= ny < len(board[0]):
                self.dfs(board, nx, ny, node, res)
        board[i][j] = ch

File: trie/212-word-search-ii/WordSearchII.java


class TrieNode {
    TrieNode[] children = new TrieNode[26];
    String word = null; // store the full word at leaf node
}

public class Solution {
    private TrieNode buildTrie(String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null)
                    node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.word = word;
        }
        return root;
    }

    public List<String> findWords(char[][] board, String[] words) {
        List<String> res = new ArrayList<>();
        TrieNode root = buildTrie(words);
        int rows = board.length, cols = board[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                dfs(board, i, j, root, res);
            }
        }
        return res;
    }

    private void dfs(char[][] board, int i, int j, TrieNode node, List<String> res) {
        char c = board[i][j];
        if (c == '#' || node.children[c - 'a'] == null) return;

        node = node.children[c - 'a'];
        if (node.word != null) {
            res.add(node.word);
            node.word = null; // prevent duplicate word
        }

        board[i][j] = '#'; // mark visited
        int[][] dirs = {{1,0}, {-1,0}, {0,1}, {0,-1}};
        for (int[] d : dirs) {
            int x = i + d[0], y = j + d[1];
            if (x >= 0 && y >= 0 && x < board.length && y < board[0].length)
                dfs(board, x, y, node, res);
        }
        board[i][j] = c; // backtrack
    }
}